controllers/integration/integration_controller.go

```diff
@@ -0,0 +1,233 @@
+		}
+		if result.CancelRequest {
+			return ctrl.Result{}, nil
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupController creates a new Integration reconciler and adds it to the Manager.
+func SetupController(manager ctrl.Manager, log *logr.Logger) error {
+	return setupControllerWithManager(manager, NewIntegrationReconciler(manager.GetClient(), log, manager.GetScheme()))
+}
+
+// setupApplicationComponentCache adds a new index field to be able to search Components by application.
+func setupApplicationComponentCache(mgr ctrl.Manager) error {
+	applicationComponentIndexFunc := func(obj client.Object) []string {
+		return []string{obj.(*hasv1alpha1.Component).Spec.Application}
+	}
+
+	return mgr.GetCache().IndexField(context.Background(), &hasv1alpha1.Component{},
+		"spec.application", applicationComponentIndexFunc)
+}
+
+// setupReleaseLinkCache adds a new index field to be able to search ReleaseLinks by application.
+func setupReleaseLinkCache(mgr ctrl.Manager) error {
+	releaseLinkIndexFunc := func(obj client.Object) []string {
+		return []string{obj.(*releasev1alpha1.ReleaseLink).Spec.Application}
+	}
+
+	return mgr.GetCache().IndexField(context.Background(), &releasev1alpha1.ReleaseLink{},
+		"spec.application", releaseLinkIndexFunc)
+}
+
+// setupReleaseCache adds a new index field to be able to search Releases by ApplicationSnapshot.
+func setupReleaseCache(mgr ctrl.Manager) error {
+	releaseIndexFunc := func(obj client.Object) []string {
+		return []string{obj.(*releasev1alpha1.Release).Spec.ApplicationSnapshot}
+	}
+
+	return mgr.GetCache().IndexField(context.Background(), &releasev1alpha1.Release{},
+		"spec.applicationSnapshot", releaseIndexFunc)
+}
+
+// setupApplicationSnapshotCache adds a new index field to be able to search ApplicationSnapshots by Application.
+func setupApplicationSnapshotCache(mgr ctrl.Manager) error {
+	applicationSnapshotIndexFunc := func(obj client.Object) []string {
+		return []string{obj.(*releasev1alpha1.ApplicationSnapshot).Spec.Application}
+	}
+
+	return mgr.GetCache().IndexField(context.Background(), &releasev1alpha1.ApplicationSnapshot{},
+		"spec.application", applicationSnapshotIndexFunc)
+}
+
+// setupControllerWithManager sets up the controller with the Manager which monitors new build PipelineRuns and filters
+// out status updates.
+func setupControllerWithManager(manager ctrl.Manager, reconciler *Reconciler) error {
+	err := setupApplicationComponentCache(manager)
+	if err != nil {
+		return err
+	}
+	err = setupReleaseLinkCache(manager)
+	if err != nil {
+		return err
+	}
+	err = setupReleaseCache(manager)
+	if err != nil {
+		return err
+	}
+	err = setupApplicationSnapshotCache(manager)
+	if err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(manager).
+		For(&tektonv1beta1.PipelineRun{}, builder.WithPredicates(tekton.BuildPipelineRunSucceededPredicate())).
+		Watches(&source.Kind{Type: &tektonv1beta1.PipelineRun{}}, &libhandler.EnqueueRequestForAnnotation{
```

Comment:
You don't need this line. This will listen for every single PipelineRun owned by an Integration CRD, which I guess it's not something you want. At least I've not seen anything in this PR doing so.

Reply:
Ack, removed it

Reply:
Ping me then. Depending where those PipelineRuns get created you will use a different approach.

Reply:
Ack, will do, thanks